### PR TITLE
Support async delete 241

### PR DESCRIPTION
**Pull request by Incarnation-p-lee**

### CODE REVIEW

File: src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java [hunk 1]
```diff
@@ -83,23 +81,44 @@ private AsyncDocumentClient getAsyncDocumentClient() {
...
+    private void validate(String collectionName) {
+        Assert.hasText(collectionName, "collection should not be null, empty or only whitespaces.");
+    }
+
+    private void validate(DocumentQuery query, String collectionName, Class<?> entityClass,
+                          List<String> partitionKeyNames) {
+        Assert.notNull(query, "DocumentQuery should not be null.");
+        Assert.hasText(collectionName, "collection should not be null, empty or only whitespaces.");
+        Assert.notNull(entityClass, "domainClass should not be null.");
+        Assert.notNull(partitionKeyNames, "partitionKeyNames should not be null.");
```
Comment (sophiaso) [inline src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java]: Can we reuse the third  validate here?

Comment (Incarnation-p-lee) [reply src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java]: yeah, make sense. I will apply the validate in another PR and reuse the implementation.

File: src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java [hunk 2]
```diff
@@ -443,21 +453,51 @@ private RequestOptions getRequestOptions(PartitionKey key, Integer requestUnit)
...
+
+    @Override
+    public <T> Observable<Boolean> existsAsync(@NonNull DocumentQuery query, @NonNull String collectionName,
+                                               @NonNull Class<T> entityClass, @NonNull List<String> partitionKeyNames) {
+        validate(query, collectionName, entityClass, partitionKeyNames);
+
+        return findAsync(query, collectionName, entityClass, partitionKeyNames)
+                .count()
+                .map(c -> c > 0)
+                .single();
```
Comment (sophiaso) [inline src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java]: Can we use the `countAsync` method to check exists? `Find` might be more heavier than count only, not confirmed though.

File: src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbOperations.java [hunk 3]
```diff
@@ -38,11 +38,14 @@
 
     void deleteCollection(String collectionName);
 
-    <T> List<T> delete(DocumentQuery query, Class<T> entityClass, String collectionName);
+    <T> List<T> delete(DocumentQuery query, String collectionName, Class<T> entityClass,
```
Comment (warrenzhu25) [inline src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbOperations.java]: I think these public should have some comments. For example, it's unclear here why I need `partitionKeyNames` for deletion

Comment (Incarnation-p-lee) [reply src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbOperations.java]: Yes, we will add comments for this interface in another PR. As the legacy code has no description here.

**Pull request merged**